components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java

```diff
@@ -0,0 +1,223 @@
+            responseHeader.setValue("Missing required parameters");
+            tokReqMsgCtx.addProperty("RESPONSE_HEADERS", new ResponseHeader[]{responseHeader});
+
+            return false;
+        }
+
+        if (log.isDebugEnabled()) {
+            log.debug("Access token validation success.");
+        }
+
+        User authorizedUser = User.getUserFromUserName(validationResponseDTO.getAuthorizedUser());
+
+        User associatedUser = new User();
+        associatedUser.setUserName(username);
+        associatedUser.setUserStoreDomain(userstoreDomain);
+        associatedUser.setTenantDomain(tenantDomain);
+
+        boolean isValidAssociation = false;
+        try {
+            isValidAssociation = validateAssociation(authorizedUser, associatedUser);
+        } catch (UserAccountAssociationException e) {
+            if (log.isDebugEnabled()) {
+                log.debug("Error while validating associations", e);
+            }
+        }
+
+        if (!isValidAssociation) {
+            if (log.isDebugEnabled()) {
+                log.debug("Authorized user: " + authorizedUser.toFullQualifiedUsername() + " not associated to: " +
+                        associatedUser.toFullQualifiedUsername());
+            }
+            ResponseHeader responseHeader = new ResponseHeader();
+            responseHeader.setKey("error-description");
+            responseHeader.setValue("Associated user is invalid.");
+            tokReqMsgCtx.addProperty("RESPONSE_HEADERS", new ResponseHeader[]{responseHeader});
+
+            return false;
+        }
+
+        tokReqMsgCtx.setAuthorizedUser(
+                OAuth2Util.getUserFromUserName(associatedUser.toFullQualifiedUsername()));
+
+        String[] allowedScopes =  getAllowedScopes(validationResponseDTO.getScope(),
+                tokReqMsgCtx.getOauth2AccessTokenReqDTO().getScope());
+        tokReqMsgCtx.setScope(allowedScopes);
+
+        if (log.isDebugEnabled()) {
+            log.debug("Issuing an access token for associated user: " + associatedUser + " with scopes: " +
+                    tokReqMsgCtx.getScope());
+        }
+
+        return true;
+    }
+
+    private String extractParameter(String param, OAuthTokenReqMessageContext tokReqMsgCtx) {
+
+        RequestParameter[] parameters = tokReqMsgCtx.getOauth2AccessTokenReqDTO().getRequestParameters();
+
+        for (RequestParameter parameter : parameters) {
+            if (param.equals(parameter.getKey())) {
+                if (parameter.getValue() != null && parameter.getValue().length > 0) {
+                    return parameter.getValue()[0];
+                }
+            }
+        }
+
+        return null;
+    }
+
+    private String[] getAllowedScopes(String[] authorizedScoped, String[] requestedScopes) {
+
+        HashSet<String> allowedScopesSet = new HashSet<>();
+        allowedScopesSet.addAll(Arrays.asList(authorizedScoped));
+        allowedScopesSet.retainAll(Arrays.asList(requestedScopes));
```

Comment:
possible null pointer Arrays.asList(null), can be avoided with ArrayUtils.nullToEmpty()